ProcessTree.cpp

```diff
@@ -0,0 +1,530 @@
+        if (recv(_proc_socket, &message, sizeof(message), 0) <= 0) {
+            Logger::Error("Error receiving from netlink socket for process monitoring");
+        }
+
+        switch (message.event.what) {
+            case proc_event::what::PROC_EVENT_FORK:
+                _processTree->AddPnForkQueue(message.event.event_data.fork.child_pid, message.event.event_data.fork.parent_pid);
+                break;
+            case proc_event::what::PROC_EVENT_EXEC:
+                _processTree->AddPnExecQueue(message.event.event_data.exec.process_pid);
+                break;
+            case proc_event::what::PROC_EVENT_EXIT:
+                _processTree->AddPnExitQueue(message.event.event_data.exit.process_pid);
+                break;
+        }
+    }
+}
+
+void ProcessTree::AddPnForkQueue(int pid, int ppid)
+{
+    struct ProcessQueueItem p = {ProcessQueueFork, ProcessTreeSource_pnotify, pid, ppid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExecQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExec, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExitQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExit, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::run()
+{
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    while (!IsStopping()) {
+        _queue_data.wait(queue_pop_lock, [&]{return _queue_data_ready;});
+        while (!_PnQueue.empty()) {
+            struct ProcessQueueItem p = _PnQueue.front();
+            _PnQueue.pop();
+            switch (p.type) {
+                case ProcessQueueFork:
+                    AddPid(p.pid, p.ppid);
+                    break;
+                case ProcessQueueExec:
+                    AddPid(p.pid);
+                    break;
+                case ProcessQueueExit:
+                    RemovePid(p.pid);
+                    break;
+                default:
+                    Logger::Error("Invalid ProcessQueueType");
+            }
+        }
+
+        // Check if it's time for routine pruning of stale pids
+        std::chrono::duration<double> elapsed_seconds = std::chrono::system_clock::now() - _last_clean_time;
+        if (elapsed_seconds.count() > CLEAN_PROCESS_INTERVAL) {
```

Comment:
Is it ok if elapsed_seconds significantly exceeds CLEAN_PROCESS_INTERVAL? If not, then _queue_data.wait() needs to be replaced with _queue_data.wait_for() and add the timeout case.

Reply:
It is okay. The clean() only needs to be called if new processes are being added and/or removed. We should be notified of these events, and these will cause the wait() to exit and the time check to happen.